File: src/main/java/kirjanpito/ui/javafx/cells/AmountTableCell.java
package kirjanpito.ui.javafx.cells;

import javafx.scene.control.*;
import javafx.scene.input.KeyCode;
import kirjanpito.ui.javafx.EntryRowModel;

import java.math.BigDecimal;
import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.text.ParseException;
import java.util.Locale;

/**
 * Editoitava taulukkosolu rahasummille.
 * Tukee suomalaista numeromuotoa (pilkku desimaalierottimena).
 */
public class AmountTableCell extends TableCell<EntryRowModel, BigDecimal> {
    
    private TextField textField;
    private final boolean isDebit;
    private static final DecimalFormat DISPLAY_FORMAT = new DecimalFormat("#,##0.00");
    private static final DecimalFormat PARSE_FORMAT;
    
    static {
        DecimalFormatSymbols symbols = new DecimalFormatSymbols(new Locale("fi", "FI"));
        PARSE_FORMAT = new DecimalFormat("#,##0.00", symbols);
        PARSE_FORMAT.setParseBigDecimal(true);
    }
    
    public AmountTableCell(boolean isDebit) {
        this.isDebit = isDebit;
        
        // Style for alignment
        setStyle("-fx-alignment: CENTER-RIGHT;");
    }
    
    @Override
    public void startEdit() {
        if (!isEmpty()) {
            super.startEdit();
            createTextField();
            setText(null);
            setGraphic(textField);
            textField.requestFocus();
            textField.selectAll();
        }
    }
    
    @Override
    public void cancelEdit() {
        super.cancelEdit();
        setText(getDisplayText(getItem()));
        setGraphic(null);
    }
    
    @Override
    public void updateItem(BigDecimal item, boolean empty) {
        super.updateItem(item, empty);
        
        if (empty) {
            setText(null);
            setGraphic(null);
        } else {
            if (isEditing()) {
                if (textField != null) {
                    textField.setText(getEditText(item));
                }
                setText(null);
                setGraphic(textField);
            } else {
                setText(getDisplayText(item));
                setGraphic(null);
                
                // Color coding
                if (item != null && item.compareTo(BigDecimal.ZERO) != 0) {
                    if (isDebit) {
                        setStyle("-fx-alignment: CENTER-RIGHT; -fx-text-fill: #059669;"); // Green
                    } else {
                        setStyle("-fx-alignment: CENTER-RIGHT; -fx-text-fill: #dc2626;"); // Red
                    }
                } else {
                    setStyle("-fx-alignment: CENTER-RIGHT; -fx-text-fill: #94a3b8;"); // Muted
                }
            }
        }
    }
    
    private void createTextField() {
        textField = new TextField(getEditText(getItem()));
        textField.setStyle("-fx-alignment: CENTER-RIGHT;");
        
        // Commit on Enter
        textField.setOnAction(e -> commitAmount());
        
        // Keyboard handling
        textField.setOnKeyPressed(e -> {
            if (e.getCode() == KeyCode.ESCAPE) {
                cancelEdit();
            } else if (e.getCode() == KeyCode.TAB) {
                commitAmount();
            }
        });
        
        // Focus lost = commit
        textField.focusedProperty().addListener((obs, wasFocused, isNowFocused) -> {
            if (!isNowFocused && isEditing()) {
                commitAmount();
            }
        });
    }
    
    private void commitAmount() {
        BigDecimal value = parseAmount(textField.getText());
        commitEdit(value);
    }
    
    private BigDecimal parseAmount(String text) {
        if (text == null || text.trim().isEmpty()) {
            return null;
        }
        
        try {
            // Replace comma with dot for parsing, handle Finnish format
            String normalized = text.trim()
                .replace(" ", "")
                .replace(",", ".");
            
            return new BigDecimal(normalized);
        } catch (NumberFormatException e) {
            // Try Finnish locale parsing
            try {
                Number parsed = PARSE_FORMAT.parse(text.trim());
                return parsed instanceof BigDecimal ? (BigDecimal) parsed : new BigDecimal(parsed.toString());
            } catch (ParseException pe) {
                return getItem(); // Keep current value
            }
        }
    }
    
    private String getDisplayText(BigDecimal amount) {
        if (amount == null || amount.compareTo(BigDecimal.ZERO) == 0) {
            return "";
        }
        return DISPLAY_FORMAT.format(amount);
    }
    
    private String getEditText(BigDecimal amount) {
        if (amount == null || amount.compareTo(BigDecimal.ZERO) == 0) {
            return "";
        }
        return DISPLAY_FORMAT.format(amount);
    }
}

File: src/main/kotlin/kirjanpito/ui/javafx/cells/AmountTableCell.kt
package kirjanpito.ui.javafx.cells

import javafx.scene.control.*
import javafx.scene.input.KeyCode
import kirjanpito.ui.javafx.EntryRowModel
import java.math.BigDecimal
import java.text.DecimalFormat
import java.text.DecimalFormatSymbols
import java.util.Locale

/**
 * Editoitava taulukkosolu rahasummille.
 * Tukee suomalaista numeromuotoa (pilkku desimaalierottimena).
 */
class AmountTableCell(
    private val isDebit: Boolean
) : TableCell<EntryRowModel, BigDecimal>() {
    
    private var textField: TextField? = null
    
    companion object {
        private val DISPLAY_FORMAT = DecimalFormat("#,##0.00")
        private val PARSE_FORMAT = DecimalFormat("#,##0.00", DecimalFormatSymbols(Locale("fi", "FI"))).apply {
            isParseBigDecimal = true
        }
    }
    
    init {
        style = "-fx-alignment: CENTER-RIGHT;"
    }
    
    override fun startEdit() {
        if (!isEmpty) {
            super.startEdit()
            createTextField()
            text = null
            graphic = textField
            textField?.requestFocus()
            textField?.selectAll()
        }
    }
    
    override fun cancelEdit() {
        super.cancelEdit()
        text = getDisplayText(item)
        graphic = null
    }
    
    override fun updateItem(item: BigDecimal?, empty: Boolean) {
        super.updateItem(item, empty)
        
        if (empty) {
            text = null
            graphic = null
        } else {
            if (isEditing) {
                textField?.text = getEditText(item)
                text = null
                graphic = textField
            } else {
                text = getDisplayText(item)
                graphic = null
                
                // Color coding
                if (item != null && item.compareTo(BigDecimal.ZERO) != 0) {
                    style = if (isDebit) {
                        "-fx-alignment: CENTER-RIGHT; -fx-text-fill: #059669;"
                    } else {
                        "-fx-alignment: CENTER-RIGHT; -fx-text-fill: #dc2626;"
                    }
                } else {
                    style = "-fx-alignment: CENTER-RIGHT; -fx-text-fill: #94a3b8;"
                }
            }
        }
    }
    
    private fun createTextField() {
        textField = TextField(getEditText(item)).apply {
            style = "-fx-alignment: CENTER-RIGHT;"
            
            // Commit on Enter
            setOnAction { commitAmount() }
            
            // Keyboard handling
            setOnKeyPressed { e ->
                when (e.code) {
                    KeyCode.ESCAPE -> cancelEdit()
                    KeyCode.TAB -> {
                        commitAmount()
                        // Tab handling delegated to EntryTableNavigationHandler
                    }
                    KeyCode.MULTIPLY -> {
                        // Asterisk key - allow parent handler to process
                        // Don't consume - let it bubble up
                    }
                    else -> {}
                }
            }
            
            // Handle asterisk typed as character
            setOnKeyTyped { e ->
                if (e.character == "*") {
                    // Don't add asterisk to text field
                    e.consume()
                    // Commit current value and let navigation handler toggle
                    commitAmount()
                }
            }
            
            // Focus lost = commit
            focusedProperty().addListener { _, _, isNowFocused ->
                if (!isNowFocused && isEditing) {
                    commitAmount()
                }
            }
        }
    }
    
    private fun commitAmount() {
        val value = parseAmount(textField?.text)
        commitEdit(value)
    }
    
    private fun parseAmount(text: String?): BigDecimal? {
        if (text.isNullOrBlank()) {
            return null
        }
        
        return try {
            // Replace comma with dot for parsing, handle Finnish format
            val normalized = text.trim()
                .replace(" ", "")
                .replace(",", ".")
            
            BigDecimal(normalized)
        } catch (e: NumberFormatException) {
            // Try Finnish locale parsing
            try {
                val parsed = PARSE_FORMAT.parse(text.trim())
                when (parsed) {
                    is BigDecimal -> parsed
                    else -> BigDecimal(parsed.toString())
                }
            } catch (pe: Exception) {
                item // Keep current value
            }
        }
    }
    
    private fun getDisplayText(amount: BigDecimal?): String {
        return if (amount == null || amount.compareTo(BigDecimal.ZERO) == 0) {
            ""
        } else {
            DISPLAY_FORMAT.format(amount)
        }
    }
    
    private fun getEditText(amount: BigDecimal?): String {
        return if (amount == null || amount.compareTo(BigDecimal.ZERO) == 0) {
            ""
        } else {
            DISPLAY_FORMAT.format(amount)
        }
    }
}
